Option 'bail' causes failed build

<!-- Please don't delete this template because we'll close your issue -->
<!-- Before creating an issue please make sure you are using the latest version of webpack. -->

# Bug report

<!-- Please ask questions on StackOverflow or the webpack Gitter. -->
<!-- https://stackoverflow.com/questions/ask?tags=webpack -->
<!-- https://gitter.im/webpack/webpack -->
<!-- Issues which contain questions or support requests will be closed. -->

**What is the current behavior?**

Setting option [bail](https://webpack.js.org/configuration/other-options/#bail) to true will break the build for unresolved modules wrapped in try...catch
After setting bail to false bundle builds successfully

**If the current behavior is a bug, please provide the steps to reproduce.**

1. Set bail: true in config
1. Add require to unresolved module wrapped in try...catch
	```
	let packageJson = {};
	
	try {
	  packageJson = require('../../../../package.json');
	} catch (e) {
	}
	```
1. Start build as always

<!-- A great way to do this is to provide your configuration via a GitHub repository -->
<!-- The most helpful is a minimal reproduction with instructions on how to reproduce -->
<!-- Repositories with too many files or large `webpack.config.js` files are not suitable -->
<!-- Please only add small code snippets directly into this issue -->
<!-- https://gist.github.com is a good place for longer code snippets -->
<!-- If your issue is caused by a plugin or loader, please create an issue on the loader/plugin repository instead -->

**What is the expected behavior?**

There is should be only a warning about the unresolved module, bundle should be built successfully.

<!-- "It should work" is not a helpful explanation -->
<!-- Explain exactly how it should behave -->

**Other relevant information:**
webpack version: 5.0.0-rc.3 
Node.js version: 12.18.4

From my test:
```js
WARNING in ./test.js 4:18-53
Module not found: Error: Can't resolve '../../../../package.json' in '/home/user/path/to/package'
```

```bash
# echo $?
0
```

--- a/lib/Compilation.js
+++ b/lib/Compilation.js
@@ -1574,10 +1574,11 @@ BREAKING CHANGE: Asset processing hooks in Compilation has been merged into a si
 				if (err) {
 					if (dependencies.every(d => d.optional)) {
 						this.warnings.push(err);
+						return callback();
 					} else {
 						this.errors.push(err);
+						return callback(err);
 					}
-					return callback(err);
 				}
 
 				if (!newModule) {

--- a/test/configCases/parsing/optional/index.js
+++ b/test/configCases/parsing/optional/index.js
@@ -1,0 +1,11 @@
+it("should not fail for optional modules with bail", () => {
+	let error;
+	try {
+		require("./not-existing");
+	} catch (e) {
+		error = e;
+	}
+	expect(() => {
+		throw error;
+	}).toThrowError();
+});

--- a/test/configCases/parsing/optional/warnings.js
+++ b/test/configCases/parsing/optional/warnings.js
@@ -1,0 +1,7 @@
+module.exports = [
+	[
+		/Module not found/,
+		/Can't resolve '\.\/not-existing' /,
+		{ details: /not-existing\.js/ }
+	]
+];

--- a/test/configCases/parsing/optional/webpack.config.js
+++ b/test/configCases/parsing/optional/webpack.config.js
@@ -1,0 +1,4 @@
+/** @type {import("../../../../").Configuration} */
+module.exports = {
+	bail: true
+};

